docs/graphql.md

```diff
@@ -0,0 +1,102 @@
+
+## Operation Name
+
+It is common (and [recommended](https://graphql.org/learn/queries/#operation-name)) to provide an _Operation Name_ for queries. Here for example `UserWithName`:
+
+```graphql
+query UserWithComments {
+  user {
+    id
+    name
+    comments {
+      body
+    }
+  }
+}
+```
+
+The point of these are to provide an easy way for the developers, when things go wrong, to pinpoint where exactly they did so.
+
+This name is available on the server too and serves as a great distinguishing key.
+
+Span name examples:
+- `GraphQL:UserWithComments`
+- `GraphQL:UpdateUser`
+
+## Multiple endpoints
+
+An app may serve multiple GraphQL endpoints. To tell them apart we can include the path in the transaction name.
+
+- `GraphQL:UserWithComments (/api/graphql)`
+
+I'm not sure this is a common thing to do, so perhaps this could be an opt-in option, like `graphql_postfix_path: true`.
+
+## Anonymous queries
+
+An Operation Name isn't required. When one isn't provided it's hard for us to tell apart the queries.
+
+Some clients generate `id`s from hashing the contents of the query (see [apollo-tooling](https://github.com/apollographql/apollo-tooling/blob/1dfd737eaf85b89b2cfb13913342e091e3c03d18/packages/apollo-codegen-core/src/compiler/visitors/generateOperationId.ts#L5)). We could choose to do something similar.
+
+A problem with this approach is that the user of the APM UI has no way to recognise queries in the transactions list before clicking through.
+
+Another approach is to simply label them `[anonymous query]` or something similar.
+
+A problem with _that_ approach is that the contents and thereby the relevant db queries and other sub-span actions that the server might do while resolving these queries may be wildly different making it hard to provide a _true_ sample waterfall.
+ These two examples for example will look the same for the top-level GraphQL spans but will represent significantly different workloads.
+
+```
+[- anonymous graphql span --------------]
+  [- 1,000x SELECT * ---------------]
+    [- 1,000 more SELECT * -]
+
+[- anonymous graphql span --------------]
+  [- SELECT id FROM users WHERE id=? -]
+```
+
+We could consider _muting_ or ignoring all sub-spans to anonymous GraphQL queries and choose to rather show nothing than potentially wrong information.
+
+Span name examples:
+- `GraphQL:ka8kadf8233kxcsc2929384384kdkdkc8383…`
+- `GraphQL:[anonymous]`
+
+## Batching/Multiplexing queries
+
+Some clients allow batching queries (see for example [apollo-link-batch-http](https://www.apollographql.com/docs/link/links/batch-http/#gatsby-focus-wrapper)
+
+So far it makes sense to update transaction names based on the span names. Essentially, in the best case, let the transactions be named after the Operation Names.
+
+However with multiple queries per HTTP request this wont work.
+
+Combining span names with `+` could work.
```

Comment:
can you go in to more details here? 
```
query {
  collections(first: 10) {
    products {
      name
      price
      instock     // assume call to separate service 
    }
  }
}
```

The above query would trigger 10 calls from graphql endpoint to the respective product API backend and 10 more to check if the product is in stock or not. How could we represent something like this in this scenario?

Reply:
Was talking about _multiplexing_ as described here https://graphql-ruby.org/queries/multiplex.html

As I understand it, that means handling multiple queries in one HTTP request. Somewhat similar to how ndjson let's you have multiple objects per request body.

In that case a single HTTP request (represented as a transaction) will in fact include multiple queries, therefore the `+` to combine them in the transaction title.